# 생물정보학 및 실습 1 - Term Project용 Mission 3 환경 셋업

```
생물정보학 및 실습 1
서울대학교 생물정보학전공/생명과학부 2023년 1학기

이 노트북은 Creative Commons BY-SA 조건으로 재사용 가능합니다.
장혜식 <hyeshik@snu.ac.kr>, 서울대학교 생명과학부
```

이 노트북은 Google Colab을 이용해서 Term Project를 할 수 있도록 환경을 준비해 줍니다. 터미널 사용이 원활하지 않은 상태이고, 데이터를 오랫동안 보관하기 어렵기 때문에 되도록 개인용 컴퓨터에서 [Anaconda](https://www.anaconda.com/)나 [JupyterLab](https://jupyterlab.readthedocs.io/en/stable/) 등을 설치하고 사용하시는 것이 편하지만, 개인용 컴퓨터 사양이 부족한 경우에는 Google CoLab을 사용할 수도 있습니다. 개인용 컴퓨터를 쓸 때는 Windows에서는 [WSL2](https://docs.microsoft.com/en-us/windows/wsl/install-win10)를 쓰시면 리눅스 커맨드라인 환경을 쓸 수 있습니다. Google Colab은 창을 닫은 후 90분 후, 또는 창을 처음 열고 12시간 후에 환경을 닫아버리기 때문에 매번 실습할 때마다 새로 셋업이 필요합니다. `Runtime->Run All`을 실행해서 전체를 한 번 실행하시면 준비됩니다. 모든 것이 잘 안 되고 뭔가 꼬인 느낌이 강하게 들 때는 `Runtime->Factory reset runtime`을 누르고 새로 시작하시면 될 수도 있습니다.


## Google Drive 마운트

데이터 파일들을 불러오고 저장하기 위해서 Google Drive를 연결합니다. **이번엔 스크립트 파일을 편집해야 하기 때문에 Google Drive 마운트를 하지 않으면 vim등 콘솔에서 쓸 수 있는 에디터를 쓰셔야 합니다.** 화면에 나오는 주소를 눌러서 인증하고 인증 코드를 그 아래에 입력하세요. 그러면 `/contents/drive/MyDrive`에서 자신의 Drive 최상단 폴더가 보입니다. Google CoLab에서 접속하지 않고, 본인 컴퓨터 또는 연구실 워크스테이션 등 다른 컴퓨터를 쓰시는 경우에는 저장공간이 충분하므로 이 부분은 생략하고 지나갑니다.

In [4]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## 실습용 Bioconda 환경 설치하기
Google CoLab에서 접속하지 않고, 본인 컴퓨터 또는 연구실 워크스테이션 등 다른 컴퓨터를 쓰시는 경우에는 따로 실습용 [Bioconda](https://bioconda.github.io/) 환경을 구축해서 이 노트북을 그 환경 안에서 띄우시면 됩니다.

CoLab을 쓰시는 분들은 다음에 따라서 실습용 프로그램들을 쉽게 설치할 수 있도록 [Bioconda](https://bioconda.github.io/) 환경을 만듭니다. 실행하고 진행상황을 관찰하다가 에러로 보이는 것이 있으면 잘 읽어보고 해결합니다.

파이썬 모듈들은 Colab쪽의 파이썬 인터프리터를 바꿀 수 없기 때문에 Conda에 설치하더라도 사용하기 어렵습니다. 파이썬 모듈은 `pip` 명령어로 설치하시면 Colab에서 불러와서 사용할 수 있습니다.

환경 초기화에 관련된 스크립트는 [GitHub 프로젝트](https://github.com/hyeshik/colab-biolab)에 있습니다.

In [11]:
!git clone https://github.com/hyeshik/colab-biolab.git
!cd colab-biolab && bash tools/setup.sh
exec(open('colab-biolab/tools/activate_conda.py').read())

fatal: destination path 'colab-biolab' already exists and is not an empty directory.
./
./root/
./root/.bashrc.biolab
./root/.bin.priority/
./root/.bin.priority/pip
./root/.bin.priority/pip2
./root/.bin.priority/pip3
./root/.condarc
./root/.profile
./root/.tmux.conf
./root/.vimrc
PREFIX=/root/conda
Unpacking payload ...
                                                                               
Installing base environment...





Preparing transaction: - done
Executing transaction: | done
installation finished.
    You currently have a PYTHONPATH environment variable set. This may cause
    unexpected behavior when running the Python interpreter in Miniconda3.
    For best results, please verify that your PYTHONPATH only points to
    directories of packages that are compatible with the Python interpreter
    in Miniconda3: /root/conda
Activated conda environment `lab'!


설정이 끝났으면, 이제 콘솔을 열어서 터미널로 접속할 수 있는 통로를 열어줍니다.

## 시퀀싱 데이터 다운로드
첫 실습 시간에 데이터 복사를 마쳤으면 이 부분은 생략하고 지나갑니다.

이제 실습 수업에서 필요한 시퀀싱 데이터를 다운로드 받아서 설치합시다. 드라이브 공간만 충분하면 데이터는 계속 저장돼 있기 때문에 학기 전체를 통해서 딱 한 번만 하시면 됩니다. Colab에서는 작업용 데이터를 Google Drive 폴더 아래의 `binfo-datapack1`에 저장합니다. 실습용 데이터는 `/content/drive/MyDrive/binfo1-datapack1`에 풀리게 됩니다.

### (Option 1) Google CoLab에서 접속한 경우
1. [Google Drive 공유폴더](https://drive.google.com/drive/folders/1QBJednh-C7A17PFOlpZHBKyDmBvd6klh?usp=sharing)에 접속합니다.
1. 중간 제목줄의 "binfo1-datapack1 ▾"에서 ▾을 누릅니다.
1. Add shortcut to Drive를 누릅니다.
1. My Drive를 선택해서 ADD SHORTCUT 버튼을 누릅니다.
1. 다음 cell을 실행해서 링크가 제대로 들어왔는지 확인합니다. (바로 확인되지 않으면 3–5분 정도 기다렸다가 하면 됩니다.)
1. 제대로 확인되면 다음으로 넘어갑니다.


In [ ]:
!ls -al /content/drive/MyDrive/binfo1-datapack1/

total 7162852
-r-------- 1 root root 1370036258 Apr 27 14:53 CLIP-35L33G.bam
-r-------- 1 root root    3118336 Apr 27 14:53 CLIP-35L33G.bam.bai
-r-------- 1 root root       7113 May 11 11:02 CLIP-let7g.bam
-r-------- 1 root root      14561 May 11 11:03 CLIP-let7g-gene.pileup
-r-------- 1 root root    2685065 May 11 11:02 CLIP-let7g.pileup
-r-------- 1 root root  883334756 Apr 27 14:53 gencode.gtf
-r-------- 1 root root   24065406 Apr 27 15:15 read-counts.txt
-r-------- 1 root root        751 Apr 27 15:15 read-counts.txt.summary
-r-------- 1 root root 1003658801 Apr 27 14:54 RNA-control.bam
-r-------- 1 root root    2276104 Apr 27 14:53 RNA-control.bam.bai
-r-------- 1 root root 1260991122 Apr 27 14:54 RNA-siLin28a.bam
-r-------- 1 root root    2710744 Apr 27 14:54 RNA-siLin28a.bam.bai
-r-------- 1 root root  981684502 Apr 27 14:54 RNA-siLuc.bam
-r-------- 1 root root    2606104 Apr 27 14:54 RNA-siLuc.bam.bai
-r-------- 1 root root  737352902 Apr 27 14:54 RPF-siLin28a.bam
-r-------- 1 r

### (Option 2) CoLab외의 다른 환경에서 접속한 경우
본인 컴퓨터나 연구실 워크스테이션 등에서 작업하는 경우에는 다음 명령어로 다운로드 받으시면 됩니다. 첫 번째에는 wget이 접속오류를 내는 경우가 많으니, 접속에러가 발생하면 다시 한 번 실행해 보세요. `/content/drive/Mydrive` 부분은 원하는 저장 경로로 바꿔서 지정하시면 됩니다.

In [ ]:
!wget -O - --no-check-certificate https://hyeshik.qbio.io/binfo/binfo1-datapack1.tar | tar -C /content/drive/MyDrive -xf -

### 파일 체크섬 확인

받은 데이터의 MD5 체크섬을 확인합니다. 한참 걸립니다. 각 파일들의 원래 체크섬은 다음과 같습니다. 한 번 비교해 보세요.

```
140aaf30bcb9276cc716f8699f04ddd6  CLIP-35L33G.bam
f1b3336ed7e2f97d562dcc71641251bd  CLIP-35L33G.bam.bai
328883a73d507eafbf5b60bd6b906201  RNA-control.bam
02073818e2f398a73c3b76e5169de1ca  RNA-control.bam.bai
b09550d09d6c2a4ce27f0226f426fdb1  RNA-siLin28a.bam
fef112c727244060ea62d3f2564a07f6  RNA-siLin28a.bam.bai
28bbd0c47d725669340c784f1b772c01  RNA-siLuc.bam
43590fdc4d81905c0432e0d1cb8cfd5b  RNA-siLuc.bam.bai
5c08a9297307bc83259e658c4474f0cc  RPF-siLin28a.bam
a1bb3e29be412dfd7fd8d16b1b1acc4c  RPF-siLin28a.bam.bai
f2eebf50943024d0116c9cd3e744c707  RPF-siLuc.bam
dc24f69e8f571fc8be30f28ce5b84fcd  RPF-siLuc.bam.bai
```

In [ ]:
!md5sum /content/drive/MyDrive/binfo1-datapack1/*

140aaf30bcb9276cc716f8699f04ddd6  /content/drive/MyDrive/binfo1-datapack1/CLIP-35L33G.bam
f1b3336ed7e2f97d562dcc71641251bd  /content/drive/MyDrive/binfo1-datapack1/CLIP-35L33G.bam.bai
5d0265965487c33475f48869fc9df2d2  /content/drive/MyDrive/binfo1-datapack1/CLIP-let7g.bam
4ea14471c5d4bf9ab1fd91b9a5b2193c  /content/drive/MyDrive/binfo1-datapack1/CLIP-let7g-gene.pileup
a53f66a5575d4ac2d3b4c44af576d44d  /content/drive/MyDrive/binfo1-datapack1/CLIP-let7g.pileup
d54d428f5872030c4ffc43058709b896  /content/drive/MyDrive/binfo1-datapack1/gencode.gtf
2698e6c8e30005ebc0593d3b9e9b5c58  /content/drive/MyDrive/binfo1-datapack1/read-counts.txt
7c9528c08ed2f1ae6e2de599db1aa8fb  /content/drive/MyDrive/binfo1-datapack1/read-counts.txt.summary
328883a73d507eafbf5b60bd6b906201  /content/drive/MyDrive/binfo1-datapack1/RNA-control.bam
02073818e2f398a73c3b76e5169de1ca  /content/drive/MyDrive/binfo1-datapack1/RNA-control.bam.bai
^C


In [ ]:
!ls -al /content/drive/MyDrive/binfo*

lrw------- 1 root root    0 Apr 28 01:58 /content/drive/MyDrive/binfo1-datapack1 -> /content/drive/.shortcut-targets-by-id/1QBJednh-C7A17PFOlpZHBKyDmBvd6klh/binfo1-datapack1

/content/drive/MyDrive/binfo1-work:
total 6744256
-rw------- 1 root root  268435456 May 18 07:47 CLIP-35L33G.bam
-rw------- 1 root root    3118336 Apr 28 02:16 CLIP-35L33G.bam.bai
-rw------- 1 root root       1185 May 18 08:26 CLIP-let7g.bam
-rw------- 1 root root          0 May 18 08:10 CLIP-let7g-gene.pileup
-rw------- 1 root root          0 May 18 08:30 CLIP-let7g.pileup
drwx------ 5 root root       4096 May 18 08:34 colab-biolab
-rw------- 1 root root  572853320 May 18 08:43 filtered-RPF-siLuc.bam
-rw------- 1 root root   28109361 May 18 08:52 fivepcounts-filtered-RPF-siLuc.txt
-rw------- 1 root root   45957833 May 18 08:48 fivepcounts-RPF-siLuc.bed
-rw------- 1 root root     855103 May 18 08:42 gencode-exons-containing-startcodon.bed
-rw------- 1 root root  883334756 Apr 28 02:16 gencode.gtf
-rw------- 1 root

# 파일 복사하기
첫 시간에 파일을 복사해 두었으면 이 부분은 생략하고 다음으로 넘어갑니다.

공유한 폴더를 나의 작업 폴더로 복사합니다.

In [ ]:
%cd /content/drive/MyDrive/binfo1-work
!cp ../binfo1-datapack1/* .

/content/drive/MyDrive/binfo1-work


# 프로그램 설치
데이터 준비가 끝났으면, 이제 이번 시간 실습에 필요한 프로그램들 몇 가지를 추가로 설치합니다.

In [12]:
!conda install -y bedtools bioawk samtools

Solving environment: / - \ | / - \ | / - \ | / - \ | / - \ | / - \ | / - \ | / - \ | / - \ | done

## Package Plan ##

  environment location: /root/conda

  added / updated specs:
    - bedtools
    - bioawk
    - samtools


The following packages will be UPDATED:

  ca-certificates    pkgs/main::ca-certificates-2023.01.10~ --> conda-forge::ca-certificates-2023.5.7-hbcca054_0 
  certifi            pkgs/main/linux-64::certifi-2022.12.7~ --> conda-forge/noarch::certifi-2023.5.7-pyhd8ed1ab_0 

The following packages will be SUPERSEDED by a higher-priority channel:

  conda              pkgs/main::conda-23.3.1-py310h06a4308~ --> conda-forge::conda-23.3.1-py310hff52083_0 




Preparing transaction: - done
Verifying transaction: | done
Executing transaction: - \ | / - \ | / - \ | / - \ | / - \ | / - done


## Mission 3

이번 미션에서는 Figure S2A 아랫쪽에 표시되는 clustered error 분석 (CIMS; Crosslinking-induced mutation sites)을 합니다. 자외선에 의해서 단백질이 붙어있다가 단백질 분해효소로 단백질을 뜯어내고 남은 자리에서는 reverse-transcriptase가 오류를 자주 냅니다. 오류율을 분석해서 단백질이 앉아 있었던 자리를 분석해 봅시다.

데이터는 지난 번과 동일하기 때문에 이미 다운로드 받으셨다면 다시 다운로드 받을 필요는 없습니다.

In [13]:
%cd /content/drive/MyDrive/binfo1-work/

/content/drive/MyDrive/binfo1-work


Genome 전체로 계산하는 방법은 자원도 꽤 쓰고 시간이 걸리기 때문에 여기서는 유전자 단위로 간략하게 하도록 합시다. 예시로 Mirlet7g를 해 보고, 추가로 Mirlet7f-1과 Mirlet7d도 봐서 비교해 봅시다.

논문은 예전 레퍼런스 유전체인 mm9을 쓰고 있기 때문에, 현재 사용되는 레퍼런스 mm39와는 좌표계에 차이가 있습니다. 새로 annotation에서 위치를 찾습니다.

In [14]:
!grep -i mirlet7g gencode.gtf

chr9	ENSEMBL	gene	106056039	106056126	.	+	.	gene_id "ENSMUSG00000065440.3"; gene_type "miRNA"; gene_name "Mirlet7g"; level 3; mgi_id "MGI:2676800";
chr9	ENSEMBL	transcript	106056039	106056126	.	+	.	gene_id "ENSMUSG00000065440.3"; transcript_id "ENSMUST00000083506.3"; gene_type "miRNA"; gene_name "Mirlet7g"; transcript_type "miRNA"; transcript_name "Mirlet7g-201"; level 3; transcript_support_level "NA"; mgi_id "MGI:2676800"; tag "basic";
chr9	ENSEMBL	exon	106056039	106056126	.	+	.	gene_id "ENSMUSG00000065440.3"; transcript_id "ENSMUST00000083506.3"; gene_type "miRNA"; gene_name "Mirlet7g"; transcript_type "miRNA"; transcript_name "Mirlet7g-201"; exon_number 1; exon_id "ENSMUSE00000522665.2"; level 3; transcript_support_level "NA"; mgi_id "MGI:2676800"; tag "basic";


**mirlet7g 유전자의 좌표를 gtf파일에서 찾음 -> chr9:106056039-10605612.**

**let-7g는 mirlet7g 유전자에서 생산되는 microRNA**

Splicing 되지 않는 transcript이기 때문에 결과가 단조롭습니다. 해당 좌표 범위로 bam을 솎아냅니다.

In [15]:
!samtools view -b -o CLIP-let7g.bam CLIP-35L33G.bam chr9:106056039-106056126
!samtools view CLIP-let7g.bam | wc -l

163


**CLIP-seq에서 얻은 CLIP-35L33G.bam에서 특정 좌표(chr9:106056039-106056126, mirlet7g 유전자)에 매핑된 read(=let-7g)를 얻고, CLIP-let7g.bam에 저장**

**해당 위치에서 let-7g는 총 163개 나옴**

각 자리에 mapping된 read들의 분포를 요약하기 위해 samtools mpileup을 이용합니다. Shannon entropy를 구할 때는 reference를 따로 줄 필요는 없습니다. samtools를 쓰지 않고 직접 read를 분석하면 더욱 더 재미있습니다.

**samtools mpileup : BAM으로부터 per-base pileup을 생성하는 도구.
이 pileup은 각 base 위치에 대해 모든 매핑된 read들의 수와 종류를 포함한다. 즉, 이 도구를 통해 각 reference 위치에 대해 모든 매핑된 read들의 정보를 알 수 있다.**

In [16]:
!samtools mpileup CLIP-let7g.bam > CLIP-let7g.pileup
!wc -l CLIP-let7g.pileup

[mpileup] 1 samples in 1 input files
<mpileup> Set max per-file depth to 8000
68548 CLIP-let7g.pileup


**결과 출력을 보면 총 68548개의 line이 있으며, 이는 이 BAM 파일 내에서 pileup 정보가 있는 위치의 총 수를 나타낸다.**

**다시 말하면, 하나의 read에 있는 각각의 base 하나하나 당 line이 생성되는 것**

Mirlet7g는 100bp도 안 되는 짧은 gene인데 `mpileup` 결과는 엄청 길게 나옵니다. 왜 그럴까요? (잘 한 번 생각해 봅시다) `인라인 코드`

이제 Mirlet7g의 핵심부분을 살펴봐서 어떻게 하면 좋을지 견적을 내 봅니다.

In [17]:
!head CLIP-let7g.pileup

chr9	106007092	N	9	^Ga^Ia^Ha^Ia^Ga^Ga^Ha^Ia^Ia	<IFIGGHII
chr9	106007093	N	9	ggggggggg	AFGIEGDII
chr9	106007094	N	9	ccccccccc	EHHIGAFEH
chr9	106007095	N	9	aaaaaaaaa	?HDIFA>II
chr9	106007096	N	9	aaaaaaaaa	EHHIG@CHH
chr9	106007097	N	9	ttttttttt	=HGEGDFII
chr9	106007098	N	9	aaaaaaaaa	?HHIGDEII
chr9	106007099	N	9	g-2nng-2nng-2nng-2nng-2nng-2nng-2nng-2nng-2nn	?IEHGEDII
chr9	106007100	N	9	*********	8IBIBC@EI
chr9	106007101	N	9	*********	8IBIBC@EI


**CLIP-let7g.pileup 파일의 처음 10줄을 출력. 각 줄은 특정 염색체 위치에서의 sequencing read 정보를 포함**

**read는 106056039 ~ 106056126 범위에 있으면서도 범위 밖으로 걸쳐있을 수 있기 때문에 106056039 ~ 106056126 밖의 위와같은 결과가 나오는 것.**

In [18]:
!awk '$2 >= 106056039 && $2 <= 106056126 { print $0; }' CLIP-let7g.pileup > CLIP-let7g-gene.pileup
!tail CLIP-let7g-gene.pileup

chr9	106056117	N	138	<<<<<<<<<CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC	BHEIG?DIIHGHII:;HGIIDGGGIIBGEB?>H<GGIIIIDIIGFHIHH9IGG=GGHIDH?DGIBIIHHGGIEIGI8GIDHDG.GGHEDIIDIDGDHIDGIFGHG;DCDDHEE@I?CGG:IHGIBGIHIIDG@DHIGG
chr9	106056118	N	139	<<<<<<<<<CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC	BHEIG?DIIHGHII3AHIIIDGGGIIB3EHFBH>GIIIIIIIIIHIIHHAIIGBDGHIDH?GIIGIIHHDIIHIGIGGI=BGG9GIHHHIIGI=IAIIGGIEHGIEBG@DHDE;IBDGG4IGGIGGIIIIGGGDGIH1G
chr9	106056119	N	138	<<<<<<<<<TTTTTTTTTTTTTTTTTTTTTTTTTT$TTTT$TTTTTTTTTTTTTTT$TTTT$TTTTTTT$TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT	BHEIG?DIIGEFII:/GIIEDDHDIIF>BH3<HBEGIHIGFIIIBIIBE:HHGD>EGGDH7BIIGGIHHDHIHDEIGEIDHGG;GIHGHHIGD:IGHD:GGEHH?3G87HD</I=DGG=IHHIHBIFII@<HDHHH1G
chr9	106056120	N	133	<<<<<<<<<TTTTTTTTTTTTTTTTTTTTTTTTTTTT$TTTT$TTT$TTTTTTTTTT$TTTTTTTT$TTTTTTTT

**pileup 파일에서 다시 특정 범위 내의 위치 (106056039와 106056126 사이)에 해당하는 줄만을 선택하여 새로운 pileup 파일에 저장**

이제 pandas에 불러와서 분석해 봅니다. R을 쓰시는 분들은 R로 보내서 작업하셔도 됩니다.

In [19]:
import pandas as pd

pileup = pd.read_csv('CLIP-let7g-gene.pileup', sep='\t', names=['chrom', 'pos', '_ref', 'count', 'basereads', 'quals'])
pileup.tail()

chrom        pos _ref  count  \
83  chr9  106056122    N     88   
84  chr9  106056123    N     31   
85  chr9  106056124    N     31   
86  chr9  106056125    N     31   
87  chr9  106056126    N     30   

                                            basereads  \
83  <<<<<<<<<C$C$C$C$C$C$C$C$C$C$C$C$C$CCCCCC$C$C$...   
84                    <<<<<<<<<CCCCCCCCCCCCCCCCCCCCCC   
85                    <<<<<<<<<AAAAAAAAAAAAAAAAAAAAAA   
86                    <<<<<<<<<GGGGGGGGGGGGGGGGGGGGGG   
87                     <<<<<<<<<GGGGGGGGGGGGGGGGGGGGG   

                                                quals  
83  BHEIG?DIIEEGIIC;GIHEGBIIIIB1=FII?FEIGGGHDBIG=H...  
84                    BHEIG?DIIIIII>GIGGIGGD>BIHHHIEH  
85                    BHEIG?DIIIIHIGGIGGGIG:9DDBIEGFH  
86                    BHEIG?DIIIIIIGGE@GFIGD;GIGIIFHD  
87                     BHEIG?DIIIIGHGHIGHI>G;GGGIGIHG

**pandas를 사용해 새로 만든 pileup 파일을 읽고, 열 이름을 지정하여 데이터프레임으로 변환**

여기서 pileup에서 계산에 사용할 것은 match와 substitution만입니다. 나머지에 해당하는 tag들은 제거합니다.

R을 쓰시는 분들은 R로 보내서 작업하셔도 됩니다.

In [20]:
import re
toremove = re.compile('[<>$*#^]')
pileup['matches'] = pileup['basereads'].apply(lambda x: toremove.sub('', x))

In [ ]:
pileup[['chrom', 'pos', 'matches']]

chrom        pos                                            matches
0   chr9  106056039                                                   
1   chr9  106056040                                                   
2   chr9  106056041                                                   
3   chr9  106056042                                                   
4   chr9  106056043                                                   
..   ...        ...                                                ...
83  chr9  106056122  CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC...
84  chr9  106056123                             CCCCCCCCCCCCCCCCCCCCCC
85  chr9  106056124                             AAAAAAAAAAAAAAAAAAAAAA
86  chr9  106056125                             GGGGGGGGGGGGGGGGGGGGGG
87  chr9  106056126                              GGGGGGGGGGGGGGGGGGGGG

[88 rows x 3 columns]

**pileup 데이터에서 필요하지 않은 문자들을 제거하는 과정을 수행. 'matches'라는 새로운 열에는 'basereads'에서 필요하지 않은 문자들이 제거된 결과가 저장**

**이 명령은 'chrom', 'pos', 'matches' 열만을 포함하는 데이터프레임을 출력**

In [21]:
pileup[pileup['pos'] == 106056094].iloc[0]['matches']

'GGGGGGAAAAAAAAGGGGGAAAAAAGCCGCAGGATGAGGTGATAAGGGAGGGGTGAAGGGCGGTGAAGGGGAAAAGAGAAAGAAAAATAAAGGGGGAGTGGGAGGAAGAAGAGAATA'

자 이제 데이터 준비가 대충 끝났습니다. 다음 순서로 진행해 보세요. (1-4번 단계는 R을 쓰시는 분들은 R로 보내서 작업하셔도 됩니다.)

1. 각 position별로 base수를 셉니다.
1. 각 position별로 Shannon entropy를 계산합니다.
1. 계산한 Shannon entropy를 [bedgraph format](https://genome.ucsc.edu/goldenPath/help/bedgraph.html)으로 출력합니다. 말은 복잡해도 실제로는 간단합니다. 4칸짜리를 만드시면 됩니다.
1. 결과 파일을 본인의 Google Drive에서 가져옵니다.
1. [UCSC Genome Browser](http://genome.ucsc.edu/cgi-bin/hgTracks?db=mm39&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chr9%3A106056039-106056126)에 접속해서 Genome은 mm39로 선택합니다.
1. 그래프 아랫쪽의 add custom tracks 버튼을 누릅니다.
1. Paste URLs or data 옆의 Choose File 버튼을 누르고 우리가 만든 bedgraph 파일을 업로드 합니다.
1. 그리고 이리 저리 감상해 보다가 View 메뉴의 PDF/PS 를 눌러서 인증샷을 한 번 찍습니다.
1. Mirlet7d와 Mirlet7f-1도 한 번 해 봅니다.

In [23]:
# 베이스 별 개수 세기
pileup['A'] = pileup['matches'].str.count('a') + pileup['matches'].str.count('A')
pileup['C'] = pileup['matches'].str.count('c') + pileup['matches'].str.count('C')
pileup['G'] = pileup['matches'].str.count('g') + pileup['matches'].str.count('G')
pileup['T'] = pileup['matches'].str.count('t') + pileup['matches'].str.count('T')

# 각 위치에서의 총 베이스 개수 계산
pileup['total'] = pileup['A'] + pileup['C'] + pileup['G'] + pileup['T']

# 베이스 별 비율 계산
pileup['A_ratio'] = pileup['A'] / pileup['total']
pileup['C_ratio'] = pileup['C'] / pileup['total']
pileup['G_ratio'] = pileup['G'] / pileup['total']
pileup['T_ratio'] = pileup['T'] / pileup['total']

# 결과 출력
pileup[['chrom', 'pos', 'A_ratio', 'C_ratio', 'G_ratio', 'T_ratio']]


chrom        pos  A_ratio  C_ratio  G_ratio  T_ratio
0   chr9  106056039      NaN      NaN      NaN      NaN
1   chr9  106056040      NaN      NaN      NaN      NaN
2   chr9  106056041      NaN      NaN      NaN      NaN
3   chr9  106056042      NaN      NaN      NaN      NaN
4   chr9  106056043      NaN      NaN      NaN      NaN
..   ...        ...      ...      ...      ...      ...
83  chr9  106056122      0.0      1.0      0.0      0.0
84  chr9  106056123      0.0      1.0      0.0      0.0
85  chr9  106056124      1.0      0.0      0.0      0.0
86  chr9  106056125      0.0      0.0      1.0      0.0
87  chr9  106056126      0.0      0.0      1.0      0.0

[88 rows x 6 columns]

In [24]:
import pandas as pd
import numpy as np

# Pileup 파일을 pandas DataFrame으로 불러오기
df = pd.read_csv("CLIP-let7g-gene.pileup", sep="\t", header=None)

# 각 base별로 count
df['A_count'] = df[4].str.count('A')
df['C_count'] = df[4].str.count('C')
df['G_count'] = df[4].str.count('G')
df['T_count'] = df[4].str.count('T')

# 전체 base 수
df['total_count'] = df['A_count'] + df['C_count'] + df['G_count'] + df['T_count']

# 확률 계산
df['P_A'] = df['A_count'] / df['total_count']
df['P_C'] = df['C_count'] / df['total_count']
df['P_G'] = df['G_count'] / df['total_count']
df['P_T'] = df['T_count'] / df['total_count']

# 0의 로그가 정의되지 않으므로, 확률이 0인 경우를 위해 작은 값을 더해줍니다.
epsilon = 1e-9

# Shannon entropy 계산
df['entropy'] = -(df['P_A']*np.log2(df['P_A']+epsilon) +
                  df['P_C']*np.log2(df['P_C']+epsilon) +
                  df['P_G']*np.log2(df['P_G']+epsilon) +
                  df['P_T']*np.log2(df['P_T']+epsilon))

# 결과 확인
print(df)


       0          1  2   3                                                  4  \
0   chr9  106056039  N   9                                          <<<<<<<<<   
1   chr9  106056040  N   9                                          <<<<<<<<<   
2   chr9  106056041  N   9                                          <<<<<<<<<   
3   chr9  106056042  N   9                                          <<<<<<<<<   
4   chr9  106056043  N   9                                          <<<<<<<<<   
..   ...        ... ..  ..                                                ...   
83  chr9  106056122  N  88  <<<<<<<<<C$C$C$C$C$C$C$C$C$C$C$C$C$CCCCCC$C$C$...   
84  chr9  106056123  N  31                    <<<<<<<<<CCCCCCCCCCCCCCCCCCCCCC   
85  chr9  106056124  N  31                    <<<<<<<<<AAAAAAAAAAAAAAAAAAAAAA   
86  chr9  106056125  N  31                    <<<<<<<<<GGGGGGGGGGGGGGGGGGGGGG   
87  chr9  106056126  N  30                     <<<<<<<<<GGGGGGGGGGGGGGGGGGGGG   

                           

In [27]:
df_bedgraph = df[[0, 1, 'entropy']].copy()  # explicitly create a copy of the DataFrame
df_bedgraph.loc[:, 1] = df_bedgraph.loc[:, 1] - 1  # bedgraph is 0-based
df_bedgraph.insert(2, 'end', df_bedgraph[1] + 1)
df_bedgraph.columns = ['chrom', 'chromStart', 'chromEnd', 'dataValue']

df_bedgraph.to_csv('CLIP-let7g-gene.bedgraph', sep='\t', index=False, header=False)


In [28]:
!pwd

'/content/drive/MyDrive/binfo1-work'

In [29]:
with open('CLIP-let7g-gene.bedgraph', 'r') as f:
    print(f.read())


chr9	106056038	106056039	
chr9	106056039	106056040	
chr9	106056040	106056041	
chr9	106056041	106056042	
chr9	106056042	106056043	
chr9	106056043	106056044	
chr9	106056044	106056045	0.676941866895496
chr9	106056045	106056046	-1.4426951595367387e-09
chr9	106056046	106056047	-1.4426951595367387e-09
chr9	106056047	106056048	-1.4426951595367387e-09
chr9	106056048	106056049	-1.4426951595367387e-09
chr9	106056049	106056050	-1.4426951595367387e-09
chr9	106056050	106056051	-1.4426951595367387e-09
chr9	106056051	106056052	-1.4426951595367387e-09
chr9	106056052	106056053	-1.4426951595367387e-09
chr9	106056053	106056054	-1.4426951595367387e-09
chr9	106056054	106056055	-1.4426951595367387e-09
chr9	106056055	106056056	-1.4426951595367387e-09
chr9	106056056	106056057	-1.4426951595367387e-09
chr9	106056057	106056058	-1.4426951595367387e-09
chr9	106056058	106056059	-1.4426951595367387e-09
chr9	106056059	106056060	-1.4426951595367387e-09
chr9	106056060	106056061	-1.4426951595367387e-09
chr9	106056061	10